error_handling.md

```diff
@@ -0,0 +1,88 @@
+- `or{}` block would be the Jack of all trades because it is used for two special kinds of sum types that are not related to each other and doesn't share anything.
+- It would be possible to define variables like `x := Option<string>('foo') // or ?string('foo')` (it's obviously a sum type that can be used everywhere) but this feature should only be used for return types of functions and struct fields. One solution would be to restrict this sum type.
+
+### Or interface
+Since both, `Option` and `Result` would be handled by an `or{}` block, both types could implement a compiler internal interface `Or` that implements the syntax sugar `or {...}`.
+
+### Syntax
+In most languages the questionmark `?` stands for optional values and could be used for `Option`. The equivalent operator would be the excalamtion mark `!` that could be used for `Result`.
+That means `?string` would mean `Option<string>` and `!string` would mean `Result<string>`. Having the syntax of `?` and `!` would help indicating that's a compiler feature and it wouldn't be needed to use sum types that obviously has some overhead.
+
+There is only one tricky part: `?!string` or `!?string`. What would it mean? Which ordner is needed? One solution would be to define that the order doesn't matter.
+`!?` or `?!` always results in `Result<Option<T>>` or `T | none | Error`. But V fmt could always reformat it to `!?` for example.
+> This is even a rare case. It's complexity isn't that awful.
+
+### Handling `Result<Option<T>>`
+Handling `?T` and `!T` is very easy and it will stay as it currently is:
+```v
+foo() or {...}
+```
+for both cases. But the `!T` adds an `err` variable to the `or{}` context. The downside of this is that the user would need to know which result arrives and he is dependent on a language sever to know it.
+An alternative approach would be `or{}` for `Option` and `catch{}` for `Result`. A catch block always has the `err` variable and the `or{}` block nerver has it.
+
+Handling the `!?T`/`Result<Option<T>>` with `or{}` would result in the initial behavior, we already have. For this a solution would be to extend the functionality of `match`.
+Matching the `!?T` type would always look like this:
+```v
+match foo() {
+  error {/* result is an error */} // Currently there is no "error" type but Error/IError. Maybe it would be nice to mark error as an internal type by writing it lowercase.
+  none {/* result is none */}
+  else {/* result is the wanted value */}
+}
+```
+> We can decide to use `it` or `foo() as x`
+
+## Open Issues
+Since this was discussed several times and is still an unsolved issue we would need to change it as long V is in alpha stage.
+This change also brings breaking changes no matter which syntax will be chosen.
+
+Currently there is no final syntax decision for this but some suggestions we have to choose from.
+
```

Comment:
`?string` <=> `string | IError`
`!string` <=> `string | None`

Then most of the existing functions that DO return errors, and not `none` can remain unchanged.
The ones that do return `none` (the minority) will use `!string` in this case.
The `or {}` blocks would be able to handle both, and `err` will be available only in or blocks that are handling `?string` (the existing behavior).

The very rare case, where someone needs to return a trinity type `string | IError | None` can be handled with an explicit match, instead of an `or {}` block.

Or blocks are a convenient short syntax for handling 2 code paths, not 3.

Reply:
Currently `field_name ?Type` in structs, does not work reliably, so it does not have much uses yet => it can be very easily changed to be `field_name !Type` .

Reply:
`!` makes more sense for errors from the connotation of exclamation points. Ideally we use that, but I agree that swapping them would be an easier transition. Keep in mind that this is probably the only time that we *can* make large breaking changes though.

Reply:
I explained somewhere that `?` is used for Optional types in most languages. It wouldn't fit to `string | IError` imo.